modules/@apostrophecms/form-widget/index.js

```diff
@@ -32,5 +34,48 @@ module.exports = {
         return _super(req, widgets);
       }
     };
+  },
+  apiRoutes(self) {
+    return {
+      post: {
+        upload: [
```

Comment:
With this approach the CAPTCHA etc. can be bypassed and scripts can blast files at the server all day and they will be kept forever, like in 2.x.

That's why I suggested a change to submit the whole thing via FormData, with the files as individual entries and the rest of the data in a single JSON blob in a single FormData entry. The file fields would be given names dynamically and those would be referenced in the JSON where needed.

Or, the actual files could be base64 encoded browser side and be directly present in the JSON. Less efficient but maybe simpler to grapple with.

That way the server can check CAPTCHA validity first before deciding to keep any of the files at all and insert them as attachments.

Reply:
Thanks, I think that makes sense.

At some point we should add a task to clean up attachments that never got attached to anything. That would require marking them in some way as in use even if they are utilized by something other than an apostrophe document.

Reply:
(I say that because you can still leak unused attachments while editing in apostrophe, although it's much less of a concern than the public uploading infinite files.)